File: TDOP-backend/src/main/java/tdop/entity/Opportunity.java
package tdop.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import com.fasterxml.jackson.annotation.JsonIgnore;
import tdop.entity.enums.OpportunityStatus;
import tdop.entity.enums.OpportunityType;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "opportunities")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Opportunity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(columnDefinition = "TEXT")
    private String requirements;

    @Column(columnDefinition = "TEXT")
    private String benefits;

    private String location;

    @Enumerated(EnumType.STRING)
    private OpportunityType type;

    private String category;
    private String salaryRange;
    private String tags;

    @Column(nullable = false)
    private LocalDateTime deadline;

    @Enumerated(EnumType.STRING)
    private OpportunityStatus status = OpportunityStatus.DRAFT;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private OrganizationProfile createdBy;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @JsonIgnore
    @OneToMany(mappedBy = "opportunity", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Application> applications = new ArrayList<>();

    // Extended fields for Developer 02
    private String sourceUrl;
    private String applicationUrl;

    private String workMode = "ONSITE";

    private String educationLevel;
    private String experienceLevel;
    private String fundingInfo;

    @Column(columnDefinition = "TEXT")
    private String eligibility;

    @Column(columnDefinition = "TEXT")
    private String requiredDocuments;

    private boolean verified = false;
    private LocalDateTime verifiedAt;
    private String verifiedBy;

    private boolean moderated = false;
    private LocalDateTime moderatedAt;
    private String moderatedBy;

    private LocalDateTime publishedAt;

    private boolean closingSoonNotified = false;

    private String riskScore = "LOW";

    private Long viewCount = 0L;
    private Long saveCount = 0L;
    private Long applicationCount = 0L;

    @JsonIgnore
    @OneToMany(mappedBy = "opportunity", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<OpportunityStatusHistory> statusHistory = new ArrayList<>();
}

File: TDOP-frontend/src/types/opportunity.ts
export type OpportunityType = 'internship' | 'full-time' | 'part-time' | 'freelance' | 'volunteer' | 'apprenticeship';
export type OpportunityStatus = 'open' | 'closed' | 'draft' | 'filled' | 'expired' | 'verified' | 'pending' | 'rejected';
export type ApplicationSort = 'newest' | 'oldest' | 'status';

export interface OpportunityFilter {
  search?: string;
  type?: OpportunityType;
  status?: OpportunityStatus;
  location?: string;
  minSalary?: number;
  maxSalary?: number;
  dateFrom?: string;
  dateTo?: string;
  experience?: string;
  skills?: string[];
  category?: string;
  educationLevel?: string;
  verified?: boolean;
  page?: number;
  limit?: number;
  sortBy?: ApplicationSort;
}

export interface Opportunity {
  id: string;
  title: string;
  description: string;
  company: string;
  companyId: string;
  type: OpportunityType;
  status: OpportunityStatus | string;
  location: string;
  isRemote: boolean;
  salaryMin?: number;
  salaryMax?: number;
  salaryCurrency?: string;
  salaryRange?: string;
  experienceLevel: string;
  requirements: string[] | string;
  responsibilities: string[];
  benefits: string[] | string;
  skills: string[];
  tags: string[];
  applicationDeadline: string;
  deadline?: string;
  publishedAt: string;
  updatedAt: string;
  createdAt: string;
  isVerified: boolean;
  isFeatured: boolean;
  views: number;
  viewCount?: number;
  saveCount?: number;
  applicationCount?: number;
  applicationsCount: number;
  images?: string[];
  category: string;
  educationLevel: string;
  applicationUrl?: string;
  sourceUrl?: string;
  workMode?: string;
  eligibility?: string;
  requiredDocuments?: string;
  fundingInfo?: string;
  verified?: boolean;
  verifiedAt?: string;
  moderated?: boolean;
  organizationName?: string;
  organizationId?: number;
}

export interface OpportunityCreate {
  title: string;
  description: string;
  type: OpportunityType;
  location: string;
  isRemote?: boolean;
  salaryMin?: number;
  salaryMax?: number;
  salaryCurrency?: string;
  experienceLevel: string;
  requirements: string[] | string;
  responsibilities: string[] | string;
  benefits: string[] | string;
  skills?: string[];
  tags?: string[];
  applicationDeadline?: string;
  deadline?: string;
  category?: string;
  educationLevel?: string;
  isFeatured?: boolean;
  salaryRange?: string;
  eligibility?: string;
  requiredDocuments?: string;
  workMode?: string;
  applicationUrl?: string;
}

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export type { PaginatedApiResponse } from './api';
